### ISSUE TITLE
Modules and Conflicting Assembly Dependencies 

### ISSUE THREAD

**Issue by Silvenga** [statiqdev/Statiq#592]
I'm trying to fix a bug that I found in Wyam, but I having issues with conflicting assembly dependencies. The compile time assemblies are as follows:

- My recipe uses `Wyam.Razor/1.1.0`
  - `Wyam.Razor/1.1.0` requires `Microsoft.AspNetCore.Mvc.Razor/1.1.2`
  - `Microsoft.AspNetCore.Mvc.Razor/1.1.2` requires `Microsoft.CodeAnalysis.CSharp/1.3.0`
  - `Microsoft.CodeAnalysis.CSharp/1.3.0` requires `Microsoft.CodeAnalysis.Common/1.3.0`
  - `Wyam.Razor/1.1.0` requires `Microsoft.CodeAnalysis.CSharp/2.4.0`
  - There is a conflict with `Microsoft.CodeAnalysis.CSharp/2.4.0`, but is specified directly so `Microsoft.CodeAnalysis.CSharp/2.4.0` is taken by the compiler.
- Wyam uses `Wyam.Configuration/1.1.0`
  - `Wyam.Configuration/1.1.0` requires `Microsoft.CodeAnalysis.CSharp/2.4.0`
  - `Microsoft.CodeAnalysis.CSharp/2.4.0` requires `Microsoft.CodeAnalysis.Common/2.4.0`
  - `Microsoft.CodeAnalysis.CSharp/2.4.0` does not have a conflict, so is taken by the compiler.

Two things occur during runtime:

- Wyam detects that it does not need to compile `config.wyam` as the hash is the same.
  - The runtime lazily loads Wyam.Razor's dependency of `Microsoft.AspNetCore.Mvc.Razor/1.1.2`, which in turn attempts to load `Microsoft.CodeAnalysis.CSharp/1.3.0`.
  - `Microsoft.CodeAnalysis.CSharp/1.3.0` can not be found as `Wyam.Razor` and `Wyam` only have references to `Microsoft.CodeAnalysis.Common/2.4.0`, so no assemblies can be found in the standard search locations.
  - Wyam throws a `FileNotFoundException` when this occurs.
- Wyam compiles the `config.wyam` file.
  - To do this, the runtime lazily loads `Microsoft.CodeAnalysis.CSharp/2.4.0` as defined during compile time. This succeeds.
  - Wyam then moves to execute the recipe.
  - My recipe executes the `Wyam.Razor/1.1.0` module. This module loads `Microsoft.AspNetCore.Mvc.Razor/1.1.2`, however, since `Microsoft.CodeAnalysis.CSharp/2.4.0` has been loaded into the app domain by the runtime, `Microsoft.CodeAnalysis.CSharp/2.4.0` is used over `Microsoft.CodeAnalysis.Common/1.3.0`.

Hopefully, I'm not completely wrong about this, but if I'm right, the classic solution to this is either upgrade Wyam.Razor's Razor version or add an explicit assembly redirect in the executing program's app.config. This doesn't seem to really be a practical fix though.

So my questions are:
- Assuming that Wyam will have a large and mature module community, there will be many modules with different levels of maturity. Modules will have different versions of different dependencies. It would suck if the user of Wyam needs to consider the versions of the different modules they want to use. In this case, how should Wyam handle this?
- How should this issue be handled now?

**Comment by daveaglick** [comment]
Thanks for the detailed analysis. I seem to be hitting a lot of issues like this in many of my projects. The whole .NET tooling and runtime story is a bit of a mess right now. I'm hoping some of this will resolve itself as Wyam moves to .NET Standard (with .NET Core client applications).

> It would suck if the user of Wyam needs to consider the versions of the different modules they want to use.

Totally agree. A third-party module ecosystem hasn't emerged yet, but I want Wyam to be supportive of third-party modules if they do become available. Obviously there will be some limits to that, but that's why Wyam.Common is split out from Wyam.Core - presumably the common code won't iterate as quickly. Even though I rev the version on everything for each release we should still have a high degree of backwards compatibility as long as the module only references Wyam.Common.

As for your specific problem, I don't think updating Razor is an option at the moment. I've attempted to do so once already and there were enough changes in the way Razor works that it cascaded into a large effort and I backed out until I had time to look at it closer. The Razor libraries have been iterating very quickly and at this point the API is totally different than what we're using. Wyam will have to update eventually, but I want to focus energy on the .NET Standard/Core port first.

There's one thing that I think will require further investigation here. `Microsoft.CodeAnalysis.CSharp` and `Microsoft.CodeAnalysis.Common` are two different packages/assemblies. The former references the latter. That means we'd never use `Microsoft.CodeAnalysis.CSharp/2.4.0` over `Microsoft.CodeAnalysis.Common/1.3.0` - what's probably going on is that the reference to `Microsoft.CodeAnalysis.CSharp/2.4.0` is looking for it's own reference to `Microsoft.CodeAnalysis.Common/2.4.0` and can't find it. I'm even a little surprised by this because Wyam has it's own assembly resolution handler - it'll automatically substitute the most recent version of an assembly when one with the same name is requested.

Two thoughts:
* It's possible the Wyam assembly resolver isn't working quite right and we're either not finding the requested assemblies (hence the exception) or not returning something the referencing assembly can use.
* Does the failure occur during Wyam configuration compilation or Razor compilation? If so, that's a whole other can of worms. The Razor libraries use their own instance of Roslyn to compile Razor pages and it's possible that compiler isn't getting the right assemblies.

**Comment by daveaglick** [comment]
Unfortunately, my suggestion at the moment would be to add a binding entry to the `app.config`. If that gets around your issue then go ahead and submit it as a PR. Eventually we shouldn't need any more binding entries (the .NET Core runtime has different binding behavior that's supposed to be better about mismatches), but for now I think it's enough to just get it to work.

**Comment by Silvenga** [comment]
Yep, totally understand Wyam's position. I've been running into similar issues with my projects that I manage too (lot's of plugins). I'm really hoping that .NetCore eventually finds a better solution.

> Does the failure occur during Wyam configuration compilation or Razor compilation?

_Facepalm, maybe a stack trace would have been useful._

The error is occurring within the AspNetCore dependency injection container when Wyam sets up Razor.

```
Exception during execution: System.AggregateException: One or more errors occurred. ---> System.IO.FileNotFoundException: Could not load file or assembly 'Microsoft.CodeAnalysis, Version=1.3.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified.
   at Microsoft.Extensions.DependencyInjection.MvcRazorMvcCoreBuilderExtensions.AddRazorViewEngineServices(IServiceCollection services)
   at Microsoft.Extensions.DependencyInjection.MvcRazorMvcCoreBuilderExtensions.AddRazorViewEngine(IMvcCoreBuilder builder)
   at Wyam.Razor.RazorCompiler..ctor(CompilationParameters parameters)
   at Wyam.Razor.RazorService.<>c__DisplayClass1_0.<Render>b__0(CompilationParameters _)
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at Wyam.Razor.RazorService.Render(RenderRequest request)
   at Wyam.Razor.Razor.<>c__DisplayClass16_0.<Execute>b__1(IDocument input)
   at Wyam.Common.Execution.TraceExceptionsExtensions.TraceExceptions[TResult](IExecutionContext context, IDocument document, Func`2 func)
   at Wyam.Common.Execution.TraceExceptionsExtensions.<>c__DisplayClass9_0.<Select>b__0(IDocument x)
   at System.Linq.Parallel.SelectQueryOperator`2.SelectQueryOperatorResults.GetElement(Int32 index)
   at System.Linq.Parallel.QueryResults`1.get_Item(Int32 index)
   at System.Linq.Parallel.PartitionedDataSource`1.ListContiguousIndexRangeEnumerator.MoveNext(T& currentElement, Int32& currentKey)
   at System.Linq.Parallel.PipelineSpoolingTask`2.SpoolingWork()
   at System.Linq.Parallel.SpoolingTaskBase.Work()
   at System.Linq.Parallel.QueryTask.BaseWork(Object unused)
   at System.Linq.Parallel.QueryTask.<>c.<.cctor>b__10_0(Object o)
   at System.Threading.Tasks.Task.InnerInvoke()
   at System.Threading.Tasks.Task.Execute()
   --- End of inner exception stack trace ---
   at System.Linq.Parallel.QueryTaskGroupState.QueryEnd(Boolean userInitiatedDispose)
   at System.Linq.Parallel.AsynchronousChannelMergeEnumerator`1.MoveNextSlowPath()
   at System.Linq.Parallel.AsynchronousChannelMergeEnumerator`1.MoveNext()
   at System.Linq.Parallel.QueryOpeningEnumerator`1.MoveNext()
   at System.Linq.Enumerable.WhereEnumerableIterator`1.MoveNext()
   at System.Linq.Buffer`1..ctor(IEnumerable`1 source)
   at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
   at System.Collections.Immutable.ImmutableArray.CreateRange[T](IEnumerable`1 items)
   at System.Collections.Immutable.ImmutableArray.ToImmutableArray[TSource](IEnumerable`1 items)
   at Wyam.Core.Execution.ExecutionPipeline.Execute(ExecutionContext context, IEnumerable`1 modules, ImmutableArray`1 inputDocuments)
   at Wyam.Core.Execution.ExecutionPipeline.Execute(Engine engine, Guid executionId)
   at Wyam.Core.Execution.Engine.Execute()
```

My current workaround is to just disable assembly caching, which works for me for now. ¯\\_(ツ)_/¯

**Comment by daveaglick** [comment]
Thanks for the stack trace - that'll help a bunch. It looks like it's happening during service registration so should be able to track down the code there.

**Comment by daveaglick** [comment]
This is hopefully resolved with version 2.0 since the assemblies used by modules all line up now. .NET Core also handles run-time binding problems better so I'm hopeful we won't see a repeat of binding hell - I've dealt with enough of that for three lifetimes by now.
